core/src/task.cpp

```diff
@@ -234,31 +234,39 @@ void Task::compute() {
 }
 
 moveit::core::MoveItErrorCode Task::plan(size_t max_solutions) {
-	auto impl = pimpl();
-	init();
-
-	// Print state and return success if there are solutions otherwise the input error_code
-	const auto success_or = [this](const int32_t error_code) -> int32_t {
-		if (numSolutions() > 0)
-			return moveit::core::MoveItErrorCode::SUCCESS;
-		printState();
-		explainFailure();
-		return error_code;
-	};
-	const double available_time = timeout();
-	const auto start_time = std::chrono::steady_clock::now();
-	while (canCompute() && (max_solutions == 0 || numSolutions() < max_solutions)) {
-		if (impl->preempt_requested_)
-			return success_or(moveit::core::MoveItErrorCode::PREEMPTED);
-		if (std::chrono::duration<double>(std::chrono::steady_clock::now() - start_time).count() >= available_time)
-			return success_or(moveit::core::MoveItErrorCode::TIMED_OUT);
-		compute();
-		for (const auto& cb : impl->task_cbs_)
-			cb(*this);
-		if (impl->introspection_)
-			impl->introspection_->publishTaskState();
-	};
-	return success_or(moveit::core::MoveItErrorCode::PLANNING_FAILED);
+	try {
+		auto impl = pimpl();
+		init();
+
+		// Print state and return success if there are solutions otherwise the input error_code
+		const auto success_or = [this](const int32_t error_code) -> int32_t {
+			// reset preempt flag
+			pimpl()->preempt_requested_ = false;
+
+			if (numSolutions() > 0)
+				return moveit::core::MoveItErrorCode::SUCCESS;
+			printState();
+			explainFailure();
+			return error_code;
+		};
+		const double available_time = timeout();
+		const auto start_time = std::chrono::steady_clock::now();
+		while (canCompute() && (max_solutions == 0 || numSolutions() < max_solutions)) {
+			if (impl->preempt_requested_)
+				return success_or(moveit::core::MoveItErrorCode::PREEMPTED);
+			if (std::chrono::duration<double>(std::chrono::steady_clock::now() - start_time).count() >= available_time)
+				return success_or(moveit::core::MoveItErrorCode::TIMED_OUT);
+			compute();
+			for (const auto& cb : impl->task_cbs_)
+				cb(*this);
+			if (impl->introspection_)
+				impl->introspection_->publishTaskState();
+		};
+		return success_or(moveit::core::MoveItErrorCode::PLANNING_FAILED);
+	} catch (const std::exception& e) {
+		pimpl()->preempt_requested_ = false;
+		throw;  // rethrow the original exception
```

Comment:
The try-catch-block is the wrong tool to cleanup before leaving the function scope: it is only considered if an exception is thrown, not when leaving the function via return. Please revert this.
What you need is a scope guard:
https://stackoverflow.com/questions/50182244/simple-way-to-execute-code-at-the-end-of-function-scope

Reply:
Would using this external [library](https://ricab.github.io/scope_guard/) be acceptable ? At a first glance it seems well tested and easy to use.